plugins/graphql-backend/src/service/router.ts

```diff
@@ -14,65 +14,88 @@
-import { ApolloServer } from 'apollo-server-express';
-import { createModule as createCatalogModule } from '@backstage/plugin-catalog-graphql';
-import { Config } from '@backstage/config';
+import { Module } from 'graphql-modules';
+import {
+  createGraphQLApp,
+  ResolverContext,
+} from '@backstage/plugin-graphql-common';
+import { createLoader as createCatalogLoader } from '@backstage/plugin-graphql-catalog';
 import helmet from 'helmet';
-import { makeExecutableSchema } from '@graphql-tools/schema';
+import DataLoader from 'dataloader';
+import { CatalogClient } from '@backstage/catalog-client';
+import { createYoga, Plugin, YogaServerInstance } from 'graphql-yoga';
+import { useGraphQLModules } from '@envelop/graphql-modules';
+import { useDataLoader } from '@envelop/dataloader';
+import { CompoundEntityRef } from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import { printSchema } from 'graphql';
 
 /** @public */
 export interface RouterOptions {
-  logger: Logger;
   config: Config;
+  logger: Logger;
+  modules?: Module[];
+  plugins?: Plugin[];
+  createLoader?: (
+    context: Omit<ResolverContext, 'loader'>,
+  ) => DataLoader<string, any>;
+  refToId?: (ref: CompoundEntityRef | string) => string;
 }
 
 /** @public */
-export async function createRouter(
-  options: RouterOptions,
-): Promise<express.Router> {
-  const catalogModule = await createCatalogModule(options);
-
-  const { createSchemaForApollo } = createApplication({
-    modules: [catalogModule],
-    schemaBuilder(input) {
-      return makeExecutableSchema({
-        ...input,
-        inheritResolversFromInterfaces: true,
-      });
-    },
-  });
-
-  const server = new ApolloServer({
-    schema: createSchemaForApollo(),
-    logger: options.logger,
-    introspection: true,
-  });
-
-  await server.start();
+export async function createRouter({
+  config,
+  logger,
+  modules,
+  plugins,
+  createLoader,
+  refToId,
+}: RouterOptions): Promise<express.Router> {
+  let yoga: YogaServerInstance<any, any> | null = null;
 
   const router = Router();
+  const discovery = SingleHostDiscovery.fromConfig(config);
+  const catalog = new CatalogClient({ discoveryApi: discovery });
+  const application = createGraphQLApp({ modules, logger });
 
   router.get('/health', (_, response) => {
     response.json({ status: 'ok' });
   });
 
-  const apolloMiddleware = server.getMiddleware({ path: '/' });
+  router.get('/schema', (_, response) => {
+    response.send(printSchema(application.schema));
+  });
 
   if (process.env.NODE_ENV === 'development')
     router.use(
       helmet.contentSecurityPolicy({
         directives: {
           defaultSrc: ["'self'", "'unsafe-inline'", 'http://*'],
+          scriptSrc: ["'self'", "'unsafe-inline'", 'https://*'],
+          imgSrc: ["'self'", 'https: data:'],
         },
       }),
     );
 
-  router.use(apolloMiddleware);
+  router.use((req, res, next) => {
+    if (!yoga) {
```

Comment:
We have to provide `graphqlEndpoint` to `yoga` otherwise it complains that `req.baseUrl` isn't equal to `graphqlEndpoint`.

Reply:
Ah alright, you likely want to grab the external base URL from `discovery.getExternalBaseUrl(pluginId)` instead then, from the plugin env.

Reply:
I thought about this and it requires from a user to pass pluginId, which is a little bit weird

Reply:
It's up to user and I see there is no mandatory to use `/api` as a root path for backstage backend. So it's just a convention, right?

Reply:
Yep but the discovery service implementation should always match the actual setup so that it returns the actual external base URL for the instance, regardless of what changes have been made to the setup

Reply:
Could you give an example of that?

Reply:
In `backend/src/index.ts`
```ts
apiRouter.use('/catalog-graphql', await graphql(graphqlEnv));

...

  const service = createServiceBuilder(module)
    .addRouter('/api-backend', apiRouter)
```